dali/kernels/common/simd.h

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_SIMD_H_
+#define DALI_KERNELS_COMMON_SIMD_H_
+
+#ifdef __SSE2__
+#include <emmintrin.h>
+#endif
+
+#include <cstddef>
+#include <type_traits>
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+namespace simd {
+
+#ifdef __SSE2__
+
+template <int n>
+struct float4x {
+  __m128 v[n];  // NOLINT
+};
+
+template <int n>
+struct i128x {
+  __m128i v[n];  // NOLINT
+};
+
+using float4x1 = float4x<1>;
+using float4x2 = float4x<2>;
+using float4x4 = float4x<4>;
+using i128x1 = i128x<1>;
+using i128x2 = i128x<2>;
+using i128x4 = i128x<4>;
+
+inline __m128i clamp_round(__m128 f, float lo, float hi) {
+  f = _mm_min_ps(_mm_max_ps(f, _mm_set1_ps(lo)), _mm_set1_ps(hi));
+  return _mm_cvtps_epi32(f);  // round
+}
+
+inline void store(int8_t *i8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i8), _mm_packs_epi16(sv0, sv1));
+}
+
+inline void store(uint8_t *u8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(u8), _mm_packus_epi16(sv0, sv1));
+}
+
+inline void store(int16_t *i16, i128x2 iv) {
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i16), _mm_packs_epi32(iv.v[0], iv.v[1]));
+}
+
+inline void store(uint16_t *u16, i128x2 iv) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 1);
+  __m128i one = _mm_set1_epi32(1);
+  // avoid signed saturation by shifting right and saving LSB, then reconsitituting it
+  __m128i hi0 = _mm_srl_epi32(iv.v[0], sh);
+  __m128i hi1 = _mm_srl_epi32(iv.v[1], sh);
+  __m128i lo0 = _mm_and_si128(iv.v[0], one);
+  __m128i lo1 = _mm_and_si128(iv.v[0], one);
+  __m128i lo = _mm_packs_epi32(lo0, lo1);
+  __m128i hi = _mm_packs_epi32(hi0, hi1);
+  __m128i out = _mm_or_si128(_mm_sll_epi16(hi, sh), lo);
```

Comment:
This function doesn't advertise that it can handle out-of-range arguments. The problem is that there's no easy way (in SSE2 at least) to just narrow 32 bits to 16 bits without _signed_ saturation. This will work for all 16-bit values, but no, it won't work for values outside 0..0xffff range.

Reply:
I mean, this function also doesn't advertise that it doesn't work for those ranges. I assumed that if you used instructions with saturation (I know there is no much choice there), you meant saturation.

Reply:
I've rewritten it so it properly clamps.